Add test stubs for all packages

martinmaly opened the pull request:
Once code coverage is enabled, all packages will be accounted for
(perhaps with 0 test coverage) in the code coverage runs.

duglin commented:
oh man - I hate ginko  :-)

arschles commented:
Does k8s use Ginkgo for its unit tests? Otherwise, I'd like to hear a use case or other reasoning for using Ginkgo over the stdlib `testing` package at this point.

duglin commented:
I don't think k8s uses it -or at least I've never noticed it

MHBauer commented:
in k8s, ginkgo is used for e2e tests. not unit tests as far as I can tell.

pmorie commented:
GInkgo is used in kube only for e2es. Stdlib testing is used for unit tests. I personally think ginkgo is pretty heavyweight and would prefer to keep to stdlib testing for unit tests unless there is some reason that we can't use it any longer.

martinmaly commented:
Because ginkgo is built in go, it follows that there is no such case where stdlib testing cannot be used any longer, or be found insufficient. question is one of boilerplate code, and cost.

That aside, I am not too familiar with k8s e2e tests - and I find that different projects use the term slightly differently - so I am not sure where the proposed line lies. As an example, I found ginkgo quite useful when testing a command line component which sends HTTP/JSON requests to an HTTP server. it was pretty easy in ginkgo to spin up the  HTTP server, add rules which verified the request from command line, and sent valid response back to command line. Would that be e2e test, or unit test?

pmorie commented:
Personally, I don't really understand where this gets us, and would prefer not to use Ginkgo for unit tests.  Can we punt on this?

martinmaly commented:
I see 2 options 1) close the PR and 2) change it to simple testing-based, almost no-op tests.

The benefit to 2) that I see is that golang coverage numbers get only generated if there are tests around. If there are none, the directory is invisible to coverage.

At this point I can go with either. 1) is of course less work for me.

duglin commented:
I like the idea of having an outline/template ready to go (and to nag us) so I'd prefer option 2.

arschles commented:
same for option 2

krancour commented:
+1 on option 2

martinmaly commented:
Ok. Updated the PR to option 2. While doing that I moved some interface-compliance tests that @arschles pointed out while reviewing https://github.com/kubernetes-incubator/service-catalog/pull/42

duglin commented on contrib/broker/k8s/broker_test.go:
```diff
@@ -0,0 +1,27 @@
+import (
+	"testing"
+
+	_ "github.com/kubernetes-incubator/service-catalog/contrib/broker/k8s"
```
are these lines needed? I thought the current dir is automatically picked up?
Although, this might be because you used the "main_test" package - why?  I think I'm used to the test files being in the same package as the src.

martinmaly replied:
I've come across both ways of doing tests, one that use the same package name, others do xxx_test. Being a bit of a noob still, I picked one. Anyway, it seems like if I drop the _test suffix things work quite nicely so I made the change. Please take another look.

martinmaly commented:
This is now ready for review, hopefully 3 LGTMs and merge :)

duglin commented:
LGTM

arschles commented:
LGTM
